README.md
# Yodlee API

[![Latest Stable Version](https://poser.pugx.org/fintech-systems/yodlee-php-api/v/stable?format=flat-square)](https://packagist.org/packages/fintech-systems/yodlee-api)
![GitHub](https://img.shields.io/github/license/fintech-systems/yodlee-api)

## Installation

### Install the package via composer:

```bash
composer require fintech-systems/yodlee-api
```

### Store private and public keys

Create a directory `/storage` and copy your private key to `/storage/private-key.pem`.
If you want to run the full testsuite, also copy `public-key.pem` to `/storage`.

### Set up the environment file

Copy `.env.example` file to `.env` and complete the details.

## Contents of .env

```
YODLEE_COBRAND_NAME=
YODLEE_COBRAND_LOGIN=
YODLEE_COBRAND_PASSWORD=
YODLEE_API_URL=
YODLEE_API_KEY=
YODLEE_USERNAME=
```

## Commands

List of Commands
----------------

Display API keys:

```
✗ php artisan yodlee:api-key
+-----------------------------------------------+-------------+
| key                                           | createdDate |
+-----------------------------------------------+-------------+
| 00000000-00000000-0000-0000-0000-000000000000 | 2021-05-06  |
+-----------------------------------------------+-------------+
```

Console Commands
----------------

The console commands contains a subset of the main API methods.

```
yodlee:accounts                       Fetch a list of Yodlee accounts
yodlee:api-key                        Fetch a list of Yodlee API keys
yodlee:delete-user                    Delete an existing Yodlee user
yodlee:event-subscriptions            Fetch a list of subscribed notification events
yodlee:get-user                       Fetch details about a Yodlee user
yodlee:providers                      Fetch a list of Yodlee providers
yodlee:provider-accounts              Fetch a list of Yodlee provider accounts
yodlee:register-user                  Register a new Yodlee user
yodlee:subscribe                      Subscribe to DATA_UPDATES event notifications
yodlee:transactions                   Fetch a list of Yodlee transactions for a user
yodlee:unsubscribe                    Unsubscribe from DATA_UPDATES event notifications
```

## Testing

Test examples:

```bash
vendor/bin/phpunit
vendor/bin/phpunit --testdox
vendor/bin/phpunit tests/ApiTest.php
vendor/bin/phpunit --filter it_can_generate_a_jwt_token tests/
ApiTest.php
XDEBUG_MODE=coverage ./vendor/bin/phpunit --coverage-html=tests/coverage-report
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Laravel Specific Config

To publish the config:

```
php artisan vendor:publish --tag=yodlee-config
```

## Local Development

- Ensure your `private-key.pem` and `public-key.pem` keys are stored in /storage

The event subscription callback URL will normally be `https://app_url/api/v1/event`.

If you're doing local development, add a temporary URL to the `.env` file, e.g:
`EVENT_CALLBACK_URL=my-app.eu-1.sharedwithexpose.com/api/v1/event`

The above example assumes you're using Expose. The start Expose with this URL with Laravel Valet, do this:

`expose share --subdomain=my-app --server=eu-1 http://my-app.test`

If you're testing with an existing project, then update composer.json in the existing project to require the file like so:

```
"repositories": [
        ...
        ,
        {
            "type": "path",
            "url": "../yodlee-api"
        }
    ],
```

Then update composer:

```
  composer require fintech-systems/yodlee-api:dev-main
...
  - Upgrading fintech-systems/yodlee-api (v0.0.17 => dev-main)
...
  - Removing fintech-systems/yodlee-api (v0.0.17)
  - Installing fintech-systems/yodlee-api (dev-main): Symlinking from ../yodlee-api
Generating optimized autoload files
```

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

## Need Help?

### Yodlee ERD

To get a high level overview of Yodlee's data structure, see here:

https://developer.yodlee.com/docs/api/1.1/Data_Model

### Yodlee API Reference

The Yodlee Developer's API Reference can be found here:

https://developer.yodlee.com/api-reference

I'm developing this API on my own time for a larger project but if you reach out I might be able to help or prioritize features.

eugene@fintechsystems.net
+27823096710


CHANGELOG.md
# Changelog

All notable changes to `yodlee-php-api` will be documented in this file.

## 0.0.26 - 2022-09-19

- Fix delete user check for 204 and return 0 otherwise return 1

## 0.0.25 - 2022-09-19

- Fixed deletedAccount, a JWT token needs to be generated before invoking DELETE

## 0.0.24 - 2022-09-16

- Added deleteAccount & deleteProviderAccount API calls

## 0.0.23 - 2022-09-04

- New method getNextTransactionUrl

## 0.0.22 - 2022-08-28

- Additional refinement of register user command error checking

## 0.0.21 - 2022-08-28

- Register user command must decode json before checking for errors

## 0.0.20 - 2022-08-28

- Added dataExtracts() method

## 0.0.19 - 2022-08-27

- Fix deleteUser commmand should be unregisterUser

## 0.0.18 - 2022-08-25

- Add commands to subscribe, fetch, and unsubscribe to event notifications for DATA_UPDATES

## 0.0.17 - 2022-08-25

- Refactor all tests to use Laravel's HTTP client instead of custom get
- Remove dependency of FintechSystem/ApiHelper
- Update composer dependencies
- Create event subscription callback API
- Try code coverage with PhpUnit, but abandon
- Simplify tests setup

## 0.0.15 - 2022-08-23

- Fix `LaravelApiHelpersCommand` renaming problem

## 0.0.14 - 2022-08-23

This will be the last version on packagist as yodlee-php-api, the next version will be simply know as yodlee-api. Also moving over to a version that can use Laravel's HTTP facade as this will greatly simplify testing.

- Renamed LaravelApiHelpers na ApiHelpers
- Added `EVENT_CALLBACK_URL` URL for callback in .env to aid in testing
- Added createSubscriptionNotificationEvent, getSubscribedNotificationEvents and deleteNotificationSubscription API calls and rudimentary tests
- Added API post and delete methods
- Created a callback url for subscription notification events and added to service provider
- In initialize_app, specify /storage as path to keys
- Simplified .gitignore by also renaming key files
- Added getSubscribedNotificationEvents

## 0.0.13 - 2022-08-20

- Updated tests to use `/storage` and added instructions on read about using this folder
- Changed all documentation URLs in the API to reflect new #tag location
- Move order of code around in API so that methods are alphabetical and not often used method are at the bottom
- General cleanup of comments
- Additional instructions in readme about installation and testing
- Updated composer dependencies
- Added a test to "Get All Users" which only seems to return the first user

## 0.0.11 - 2021-10-22

- changed default for GetTransactions from 30 to 90 days

## 0.0.6 - 0.0.10

- various storage issues with private key and finding it standalone versus when using Laravel

## 0.0.5 - 2021-10-17

- refactor and cleanup, remove more legacy methods especially for accounts and transactions

## 0.0.4 - 2021-10-17

- fix problem whereby get provider accounts didn't take a username

## 0.0.3 - 2021-10-17

- added new tests to register and delete users
- removed test to get accounts as 'default-user' doesn't have any accounts linked via Fastlink
- remove Log:: because it was giving facade error
- documented facade error in the README, and then split off error into an ERRORS.md

## 0.0.2 - 2021-10-16

- first working version

## 0.0.1 - 2021-10-12

- initial release as a package

## 0.0.0 - 2021-10-10

- pre-release
- removed "dd" depenency and added orchestra testbench instead
- added *.json to .gitignore
- prototype converted from jacquestrdx123


src/YodleeApi.php
<?php

namespace FintechSystems\YodleeApi;

use Carbon\Carbon;
use FintechSystems\LaravelApiHelpers\Api;
use FintechSystems\YodleeApi\Contracts\BankingProvider;
use Firebase\JWT\JWT;
use Illuminate\Support\Facades\Http;

class YodleeApi implements BankingProvider
{
    private $privateKey = '/storage/private-key.pem';

    private $cobrandName;

    private $apiUrl;

    private $apiKey;

    private $username;

    private $header;

    public function __construct(array $client)
    {
        $this->cobrandName = $client['cobrand_name'];
        $this->apiUrl = $client['api_url'];
        $this->apiKey = $client['api_key'];
        $this->username = $client['username'];

        $cwd = str_replace('/public', '', getcwd());

        $this->privateKey = file_get_contents(
            $cwd.$this->privateKey
        );

        $token = $this->generateGenericJwtToken();

        $this->header = [
            'Api-Version' => 1.1,
            'Authorization' => 'Bearer '.$token,
            'Cobrand-Name' =>  $this->cobrandName,
            'Content-Type' => 'application/json',
        ];
    }

    /**
     * Build a get command.
     *
     * When no username, build header using the default username as defined in the .env file
     */
    public function get($endpoint, $username = null)
    {
        if ($username == null) {
            $token = $this->generateJwtToken($this->username);
        } else {
            $token = $this->generateJwtToken($username);
        }

        $header = [
            'Api-Version' => 1.1,
            'Authorization' => 'Bearer '.$token,
            'Cobrand-Name' => $this->cobrandName,
            'Content-Type' => 'application/json',
        ];

        return Http::withHeaders($header)->get(
            $this->apiUrl.$endpoint,
        );
    }

    /**
     * Call a data extracts url.
     *
     * https://developer.yodlee.com/api-reference#tag/DataExtracts
     * https://developer.yodlee.com/docs/api/1.1/DataExtracts
     */
    public function dataExtracts($url)
    {
        return $this->get($url);
    }

    /**
     * Add a new user (consumer) to the system.
     *
     * https://developer.yodlee.com/api-reference/aggregation#tag/User/operation/registerUser
     */
    public function registerUser(
        $loginName,
        $email,
        $currency = 'ZAR',
        $timeZone = 'GMT+2',
        $dateFormat = 'yyyy-MM-dd',
        $locale = 'en_ZA'
    ) {
        $postFields = '
        {
            "user": {
              "loginName":"'.$loginName.'",              
              "email": "'.$email.'",
              "preferences": {                
                "currency": "'.$currency.'",
                "timeZone": "'.$timeZone.'",
                "dateFormat": "'.$dateFormat.'",
                "locale": "'.$locale.'"
              }              
            }
          }
        ';

        $url = $this->apiUrl.'user/register';

        return Http::withHeaders(
            $this->header
        )->post($url, json_decode($postFields, true));
    }

    /**
     * Delete a user (Yodlee consumer). This is also called unregister a user.
     *
     * Bearer is a JWT token of the user to be deleted
     *
     * On success this method returns an empty string but on the CURL request is generates a 204 (Success without content)
     *
     * https://developer.yodlee.com/api-reference/aggregation#tag/User/operation/unregister
     */
    public function unregisterUser($user)
    {
        $url = $this->apiUrl.'user/unregister';

        $header = [
            'Api-Version' => '1.1',
            'Authorization' => 'Bearer '.$this->generateJwtToken($user),
            'Cobrand-Name' => $this->cobrandName,
            'Content-Type' => 'application/json',
        ];

        return Http::withHeaders(
            $header
        )->delete($url);
    }

    /**
     * Configs Operation createSubscriptionNotificationEvent.
     *
     * https://developer.yodlee.com/api-reference#tag/Configs/operation/createSubscriptionNotificationEvent
     */
    public function createSubscriptionNotificationEvent($eventName, $callbackUrl = '')
    {
        if (env('YODLEE_EVENT_CALLBACK_URL')) {
            $callbackUrl = env('YODLEE_EVENT_CALLBACK_URL');
        } else {
            $callbackUrl = config('app.url').'/yodlee/event';
        }

        $data = [
            'event' => [
                'callbackUrl' => $callbackUrl,
            ],
        ];

        // $data = '{ "event": {
        //             "callbackUrl":"' . $callbackUrl . '"
        //         }}';

        $url = $this->apiUrl."/configs/notifications/events/$eventName";

        return Http::withHeaders($this->header)
            ->post($url, $data);
    }

    /**
     * getSubscribedNotificationEvents.
     *
     * https://developer.yodlee.com/api-reference#tag/Configs/operation/getSubscribedNotificationEvents
     */
    public function getSubscribedNotificationEvents()
    {
        return $this->get('/configs/notifications/events');
    }

    /**
     * deleteNotificationSubscription.
     *
     * https://developer.yodlee.com/api-reference#tag/Configs/operation/deleteSubscribedNotificationEvent
     */
    public function deleteNotificationSubscription($eventName)
    {
        $url = $this->apiUrl."/configs/notifications/events/$eventName";

        return Http::withHeaders($this->header)
            ->delete($url);
    }

    /**
     * Get all accounts of a user held at the providers, i.e. all their bank accounts.
     *
     * https://developer.yodlee.com/api-reference/aggregation#tag/Accounts/operation/getAllAccounts
     */
    public function getAccounts($user = null)
    {
        return $this->get('accounts', $user);
    }

    /**
     * Delete account.
     *
     * When deleting an account the Http client header must include a JWT token based on username
     *
     * https://developer.yodlee.com/api-reference/aggregation#tag/Accounts/operation/deleteAccount
     */
    public function deleteAccount($username, $accountId)
    {
        $url = $this->apiUrl."/accounts/$accountId";

        $token = $this->generateJwtToken($username);

        $header = [
            'Api-Version' => 1.1,
            'Authorization' => 'Bearer '.$token,
            'Cobrand-Name' => $this->cobrandName,
            'Content-Type' => 'application/json',
        ];

        return Http::withHeaders($header)
            ->delete($url);
    }

    /**
     * This endpoint provides the list of API keys that exist for a customer.
     *
     * You can use one of the following authorization methods to access this API:
     *  - cobsession
     *  - JWT token
     *
     * Not available in developer sandbox environment.
     *
     * https://developer.yodlee.com/api-reference/aggregation#tag/Auth/operation/getApiKeys
     */
    public function getApiKeys()
    {
        return $this->get('auth/apiKey');
    }

    /**
     * Get a list of all the providers (financial institutions) that have been added by Yodlee based on a country code.
     *
     * https://developer.yodlee.com/api-reference/aggregation#tag/Providers/operation/getAllProviders
     */
    public function getProviders($priority = 'cobrand', $iso = 'ZA')
    {
        return $this->get("providers?priority=$priority&countryISOCode=$iso");
    }

    /**
     * Get a list of provider (banking) accounts added by the user.
     *
     * This includes all failed and successfully added provider accounts.
     *
     * https://developer.yodlee.com/api-reference/aggregation#tag/ProviderAccounts
     */
    public function getProviderAccounts($username)
    {
        return $this->get('providerAccounts', $username);
    }

    /**
     * deleteProviderAccount.
     *
     * https://developer.yodlee.com/api-reference/aggregation#tag/ProviderAccounts/operation/deleteProviderAccount
     */
    public function deleteProviderAccount($providerAccountId)
    {
        $url = $this->apiUrl."/providerAccounts/$providerAccountId";

        return Http::withHeaders($this->header)
            ->delete($url);
    }

    /**
     * Get all transactions for the last 90 days.
     *
     * https://developer.yodlee.com/api-reference/aggregation#tag/Transactions/operation/getTransactions
     */
    public function getTransactions($username, $fromDate = null)
    {
        $fromDate == null
            ? $fromDate = Carbon::now()->subDays(90)->format('Y-m-d')
            : $fromDate = $fromDate;

        $url = "transactions?fromDate=$fromDate";

        return $this->get($url, $username);
    }

    /**
     * Get transactions for the last 90 days for a specific account only.
     *
     * https://developer.yodlee.com/api-reference/aggregation#tag/Transactions/operation/getTransactions
     */
    public function getTransactionsByAccount($username, $accountId, $fromDate = null)
    {
        $fromDate == null
            ? $fromDate = Carbon::now()->subDays(90)->format('Y-m-d')
            : $fromDate = $fromDate;

        $url = "transactions?accountId=$accountId&fromDate=$fromDate";

        return $this->get($url, $username);
    }

    /**
     * The get user details service is used to get the user profile information
     * and the application preferences set at the time of user registration.
     *
     * https://developer.yodlee.com/api-reference#tag/User/operation/getUser
     */
    public function getUser($username)
    {
        return $this->get('user', $username);
    }

    /**
     * This endpoint is used to generate an API key.
     *
     * https://developer.yodlee.com/api-reference/aggregation#tag/Auth/operation/generateApiKey
     */
    public function generateApiKey($url, $cobrandArray, $publicKey)
    {
        $curl = curl_init();

        $publicKey = preg_replace('/\n/', '', $publicKey);

        curl_setopt_array($curl, [
            CURLOPT_URL => $url,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_ENCODING => '',
            CURLOPT_MAXREDIRS => 10,
            CURLOPT_TIMEOUT => 0,
            CURLOPT_FOLLOWLOCATION => true,
            CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
            CURLOPT_CUSTOMREQUEST => 'POST',
            CURLOPT_POSTFIELDS => '{
        		"publicKey": "'.$publicKey.'"
  			}',
            CURLOPT_HTTPHEADER => [
                'Api-Version: 1.1',
                'Authorization: cobSession='.$cobrandArray['cobSession'],
                'Cobrand-Name: '.$cobrandArray['cobrandName'],
                'Content-Type: application/json',
            ],
        ]);

        $response = curl_exec($curl);

        curl_close($curl);

        $object = json_decode($response);

        // If a key doesn't exist after the API call, then just return the response which contains the error
        return $object->apiKey['0']->key ?? $response;
    }

    /**
     * Generate a JWT token for a user.
     */
    public function generateJwtToken($user)
    {
        $payload = [
            'sub' => $user,
            'iss' => $this->apiKey,
            'iat' => time(),
            'exp' => time() + 1800,
        ];

        return JWT::encode($payload, $this->privateKey, 'RS512');
    }

    /**
     * Generic tokens are used for global API calls that does not pertain to a user.
     *
     * The key format must be precise other you may get this error:
     *  openssl_sign(): Supplied key param cannot be coerced into a private key
     *
     * Yodlee documentation overview of JWT:
     * https://developer.yodlee.com/docs/overview-json-web-tokens
     * https://developer.yodlee.com/docs/api/1.1/getting-started-with-jwts
     */
    public function generateGenericJwtToken()
    {
        $payload = [
            'iss' => $this->apiKey,
            'iat' => time(),
            'exp' => time() + 1800,
        ];

        return JWT::encode($payload, $this->privateKey, 'RS512');
    }

    public function getCobSession($url, $cobrandArray)
    {
        $curl = curl_init();

        curl_setopt_array($curl, [
            CURLOPT_URL => $url,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_ENCODING => '',
            CURLOPT_MAXREDIRS => 10,
            CURLOPT_TIMEOUT => 0,
            CURLOPT_FOLLOWLOCATION => true,
            CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
            CURLOPT_CUSTOMREQUEST => 'POST',
            CURLOPT_POSTFIELDS => '{
        		"cobrand":      {
					"cobrandLogin": "'.$cobrandArray['cobrandLogin'].'",
					"cobrandPassword": "'.$cobrandArray['cobrandPassword'].'"
         		}
    		}',
            CURLOPT_HTTPHEADER => [
                'Api-Version: 1.1',
                'Cobrand-Name: '.$cobrandArray['cobrandName'],
                'Content-Type: application/json',
                'Cookie: JSESSIONID=xxx', // REDACTED TODO Research
            ],
        ]);

        $response = curl_exec($curl);

        curl_close($curl);

        $object = json_decode($response);

        return $object->session->cobSession;
    }

    /**
     * Yodlee getTransactions request which return more than 500 transactions come in two or three parts,
     * the first part has next, and count, and the second part will have previous, next and count.
     * This method strips all unused characters and returns only the relevant part of the URL.
     */
    public function getNextTransactionUrl($link, $rel = 'next')
    {
        $link = str_replace(': ', '', $link);

        $link = str_replace('/', '', $link);

        $parts = explode(',', $link);

        foreach ($parts as $part) {
            $subPart = explode(';', $part);

            if ($subPart[1] == "rel=$rel") {
                return trim($subPart[0]);
            }
        }

        return null;
    }
}
